### PR TITLE
Test(actions): add unit tests for rule actions

### PR DESCRIPTION
**Pull request by dakshhhhh16**
## Summary
This PR improves the test coverage of the rule engine actions logic by introducing comprehensive unit tests for [internal/engine/actions/actions.go](cci:7://file:///Users/dakshpathak/Desktop/Minder/minder/internal/engine/actions/actions.go:0:0-0:0). The goal is to ensure the reliability of [shouldRemediate](cci:1://file:///Users/dakshpathak/Desktop/Minder/minder/internal/engine/actions/actions.go:137:0-179:1), [shouldAlert](cci:1://file:///Users/dakshpathak/Desktop/Minder/minder/internal/engine/actions/actions.go:181:0-237:1), and metadata extraction functions. We test across all evaluation states (skipped, error, success, failure) to prevent accidental regressions in auto-remediation and alert firing behaviors. There are no new dependencies required.
## Testing
- Ran `go test ./internal/engine/actions -v`. 
- Verified all fallback paths, edge cases involving previously skipped actions, and PR remediation specific cases trigger the [On](cci:2://file:///Users/dakshpathak/Desktop/Minder/minder/internal/engine/eval/vulncheck/review.go:603:0-603:34), `Off`, and `DoNothing` commands securely.

### CODE REVIEW

File: internal/engine/actions/actions_test.go [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+	assert.Empty(t, remMeta)
+
+	var alertMeta map[string]any
+	require.NoError(t, json.Unmarshal(res.AlertMeta, &alertMeta))
+	assert.Empty(t, alertMeta)
+}
+
+func TestGetRemediationMeta(t *testing.T) {
+	t.Parallel()
+	t.Run("nil profile", func(t *testing.T) {
```
Comment (evankanderson) [inline internal/engine/actions/actions_test.go]: If you aren't using a table test (data-driven test), I don't think there's a need to use `t.Run()` here (which will avoid the warnings on `t.Parallel` usage).

Comment (evankanderson) [reply internal/engine/actions/actions_test.go]: (here and below)

File: internal/engine/actions/actions_test.go [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+
+	tests := []struct {
+		name           string
+		prevEvalFromDb *db.ListRuleEvaluationsByProfileIdRow
+		evalErr        error
+		expected       engif.ActionCmd
+	}{
+		{
+			name:           "eval success, prev skipped -> off",
+			prevEvalFromDb: &db.ListRuleEvaluationsByProfileIdRow{RemStatus: db.RemediationStatusTypesSuccess},
```
Comment (evankanderson) [inline internal/engine/actions/actions_test.go]: Given that you're only ever setting `RemStatus`, this would read more simply if you made the table data `RemediationStatusTypes`, and then created the row in the test.

```golang
			t.Parallel()
			prevRow := db.ListRuleEvaluationsByProfileIdRow{RemStatus: tt.prevEval}
			got := shouldRemediate(&prevRow, tt.evalErr)
			assert.Equal(t, tt.expected, got)
```

File: internal/engine/actions/actions_test.go [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+	assert.ErrorIs(t, res.AlertErr, enginerr.ErrActionSkipped)
+
+	var remMeta map[string]any
+	require.NoError(t, json.Unmarshal(res.RemediateMeta, &remMeta))
+	assert.Empty(t, remMeta)
+
+	var alertMeta map[string]any
+	require.NoError(t, json.Unmarshal(res.AlertMeta, &alertMeta))
+	assert.Empty(t, alertMeta)
+}
```
Comment (evankanderson) [inline internal/engine/actions/actions_test.go]: FWIW, `getDefaultResult` is already tested transitively, which seems sufficient for an internal implementation detail function with no real way to return error.  (In fact, we should probably just use a static string rather than `json.Marshal` in that case.)

File: internal/engine/actions/actions_test.go [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+
+func TestGetRemediationMeta(t *testing.T) {
```
Comment (evankanderson) [inline internal/engine/actions/actions_test.go]: IMO, `getRemediationMeta` isn't really helpful to test on its own, as it's basically a nil-check helper.  Getting coverage for the `!skipRemediate` and `!skipAlert` lines 96-110 in `actions.go` would be much more meaningful (and would incidentally cover these lines).

File: internal/engine/actions/actions_test.go (hunk 2 above)
Comment (evankanderson) [reply internal/engine/actions/actions_test.go]: Alternatively, you might want to test the case of a nil DB row (e.g. no previous evaluation result).

File: internal/engine/actions/actions_test.go [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+			name:           "eval success, prev skipped -> do nothing",
+			prevEvalFromDb: &db.ListRuleEvaluationsByProfileIdRow{RemStatus: db.RemediationStatusTypesSkipped},
+			evalErr:        nil,
+			expected:       engif.ActionCmdDoNothing,
+		},
+		{
+			name:           "eval failure, prev skipped -> on",
+			prevEvalFromDb: &db.ListRuleEvaluationsByProfileIdRow{RemStatus: db.RemediationStatusTypesSkipped},
+			evalErr:        enginerr.NewErrEvaluationFailed("failed"),
+			expected:       engif.ActionCmdOn,
```
Comment (evankanderson) [inline internal/engine/actions/actions_test.go]: Can you add a test for `eval failure, prev error` and `eval err, prev err`?  I feel like there might be something wrong in those combinations.

Comment (evankanderson) [reply internal/engine/actions/actions_test.go]: (As in, we have an actual logic bug that testing would have revealed...)

File: internal/engine/actions/actions_test.go [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+		name           string
+		prevEvalFromDb *db.ListRuleEvaluationsByProfileIdRow
+		evalErr        error
+		remErr         error
+		remType        string
+		expected       engif.ActionCmd
+	}{
+		{
+			name:           "successful non-pr remediation with alert not off",
+			prevEvalFromDb: &db.ListRuleEvaluationsByProfileIdRow{AlertStatus: db.AlertStatusTypesOn},
```
Comment (evankanderson) [inline internal/engine/actions/actions_test.go]: Ditto on this type here.

File: internal/engine/actions/actions_test.go [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+			remErr:         nil,
+			remType:        pull_request.RemediateType,
+			expected:       engif.ActionCmdDoNothing,
+		},
+		{
+			name:           "eval success and alert on -> off",
+			prevEvalFromDb: &db.ListRuleEvaluationsByProfileIdRow{AlertStatus: db.AlertStatusTypesOn},
+			evalErr:        nil,
+			remErr:         nil,
+			remType:        pull_request.RemediateType,
```
Comment (evankanderson) [inline internal/engine/actions/actions_test.go]: Does the remediation type here matter?

File: internal/engine/actions/actions_test.go [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+			expected:       engif.ActionCmdOff,
+		},
+		{
+			name:           "eval success and alert already off -> do nothing",
+			prevEvalFromDb: &db.ListRuleEvaluationsByProfileIdRow{AlertStatus: db.AlertStatusTypesOff},
+			evalErr:        nil,
+			remErr:         nil,
+			remType:        pull_request.RemediateType,
```
Comment (evankanderson) [inline internal/engine/actions/actions_test.go]: Ditto

File: internal/engine/actions/actions_test.go [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+
+	tests := []struct {
+		name           string
+		prevEvalFromDb *db.ListRuleEvaluationsByProfileIdRow
+		evalErr        error
+		remErr         error
+		remType        string
+		expected       engif.ActionCmd
+	}{
+		{
```
Comment (evankanderson) [inline internal/engine/actions/actions_test.go:148]: Slight preference for ordering things as follows:

1. "Success" / "happy path" cases - start with "rule passed", then "rule eval detected a failure"
2. "Expected errors" - e.g. "rule eval errored" (like a bad rule definition)
3. "Unexpected errors" / "edge cases" - e.g. `remErr` is non-nil.

File: internal/engine/actions/actions_test.go [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+			name:      "eval error -> do nothing",
+			prevAlert: db.AlertStatusTypesOff,
+			evalErr:   errors.New("generic error"),
+			remErr:    nil,
+			remType:   pull_request.RemediateType,
+			expected:  engif.ActionCmdDoNothing,
+		},
+		// Edge cases
+		{
+			name:      "nil DB row, eval failure -> on",
```
Comment (evankanderson) [inline internal/engine/actions/actions_test.go]: Lint says spacing here is off by one, sorry!

File: internal/engine/actions/actions_test.go [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+			name:     "eval failure, prev skipped -> on",
+			prevEval: db.RemediationStatusTypesSkipped,
+			evalErr:  enginerr.NewErrEvaluationFailed("failed"),
+			expected: engif.ActionCmdOn,
+		},
+		{
+			name:     "eval failure, prev success -> do nothing",
+			prevEval: db.RemediationStatusTypesSuccess,
+			evalErr:  enginerr.NewErrEvaluationFailed("failed"),
+			expected: engif.ActionCmdDoNothing,
```
Comment (evankanderson) [inline internal/engine/actions/actions_test.go:52]: This seems weird... shouldn't we be saying to turn on the remediation at this point?

As an aside, we're basically doing some state tracking here to figure out how to handle the remediation given previous status.  For example, if we have a PR type remediation:

* start failing -> create a PR
* still failing -> already have a PR, don't need to act
* start passing, was failing -> close the PR
* keep passing -> close the PR if one was open

I'm not sure whether we should consider changing this to be level set ("there should be a PR open") or keep this as-is, but that's a larger change.

File: internal/engine/actions/actions_test.go [hunk 12]
```diff
@@ -0,0 +1,220 @@
...
+		{
+			name:     "eval error, prev skipped -> do nothing",
+			prevEval: db.RemediationStatusTypesSkipped,
+			evalErr:  errors.New("random error"),
+			expected: engif.ActionCmdDoNothing,
+		},
+		{
+			// NOTE: EvalStatusTypesError has an empty case body in shouldRemediate,
+			// so eval errors fall through to the default DoNothing. This may be a bug
+			// (see the comment on cases Error/Success in shouldRemediate).
```
Comment (evankanderson) [inline internal/engine/actions/actions_test.go:64]: I think this is *do nothing*, because the evaluation just shrugged.  I think "eval error, prev skipped" is wrong -- can we change that to `ActionCmdOn`?

File: internal/engine/actions/actions_test.go (hunk 11 above)
Comment (dakshhhhh16) [reply internal/engine/actions/actions_test.go:52]: Yeah, it definitely looks a bit weird! The reason it's `do nothing` here is to prevent infinite remediation loops (where it remediates, passes, fails again, and then tries to remediate again). I've added a note to the test name to make that a bit clearer.

I totally agree that a proper state machine or level-setting approach would be much cleaner long-term, but probably best to leave it as-is for now so we don't balloon the scope of this PR!

File: internal/engine/actions/actions_test.go (hunk 10 above)
Comment (evankanderson) [reply internal/engine/actions/actions_test.go]: This line is still off.  Running `make lint-fix` should auto-fix it for you, assuming you've run `make bootstrap` to get the tools installed.

Comment (dakshhhhh16) [reply internal/engine/actions/actions_test.go]: I was working on the tests and, I have by mistakenly pushed some prs which are having non-test files.

Comment (dakshhhhh16) [reply internal/engine/actions/actions_test.go]: > This line is still off. Running `make lint-fix` should auto-fix it for you, assuming you've run `make bootstrap` to get the tools installed.

Fixing this

File: internal/engine/actions/actions_test.go [hunk 13]
```diff
@@ -0,0 +1,220 @@
...
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+			var prevRow *db.ListRuleEvaluationsByProfileIdRow
+			if !tt.nilRow {
+				prevRow = &db.ListRuleEvaluationsByProfileIdRow{RemStatus: tt.prevEval}
+			}
+			got := shouldRemediate(prevRow, tt.evalErr)
+			assert.Equal(t, tt.expected, got)
+		})
+	}
```
[the remark this answers is not in the document]
Comment (evankanderson) [reply internal/engine/actions/actions_test.go:113]: We're using go 1.25, so this footgun is very much locked away.  Sorry copilot.

File: internal/engine/actions/actions_test.go [hunk 14]
```diff
@@ -0,0 +1,220 @@
...
+
+	tests := []struct {
+		name      string
+		prevAlert db.AlertStatusTypes
+		nilRow    bool
+		evalErr   error
+		remErr    error
+		remType   string
+		expected  engif.ActionCmd
+	}{
```
[the remark this answers is not in the document]
Comment (evankanderson) [reply internal/engine/actions/actions_test.go:127]: (alternatively, you can remove `remErr`)

I'd noticed this but wasn't bothered by it.

**Comment by dakshhhhh16** [comment]
@evankanderson 
Any amendments into this?
